x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyBoolQueryBuilder.java

```diff
@@ -0,0 +1,219 @@
+import org.elasticsearch.xpack.core.security.authc.Authentication;
+import org.elasticsearch.xpack.security.authc.ApiKeyService;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Function;
+
+public class ApiKeyBoolQueryBuilder extends BoolQueryBuilder {
+
+    // Field names allowed at the index level
+    private static final Set<String> ALLOWED_EXACT_INDEX_FIELD_NAMES =
+        Set.of("doc_type", "name", "api_key_invalidated", "creation_time", "expiration_time");
+
+    private ApiKeyBoolQueryBuilder() {}
+
+    public static ApiKeyBoolQueryBuilder build(QueryBuilder queryBuilder, Authentication authentication) {
+        final ApiKeyBoolQueryBuilder finalQuery = new ApiKeyBoolQueryBuilder();
+        if (queryBuilder != null) {
+            QueryBuilder processedQuery = doProcess(queryBuilder);
+            if (false == processedQuery instanceof BoolQueryBuilder) {
+                finalQuery.must(processedQuery);
+            } else {
+                final BoolQueryBuilder boolQueryBuilder = (BoolQueryBuilder) processedQuery;
+                finalQuery.minimumShouldMatch(boolQueryBuilder.minimumShouldMatch());
+                finalQuery.adjustPureNegative(boolQueryBuilder.adjustPureNegative());
+                boolQueryBuilder.must().forEach(finalQuery::must);
+                boolQueryBuilder.should().forEach(finalQuery::should);
+                boolQueryBuilder.mustNot().forEach(finalQuery::mustNot);
+                boolQueryBuilder.filter().forEach(finalQuery::filter);
+            }
+        }
+        finalQuery.filter(QueryBuilders.termQuery("doc_type", "api_key"));
+
+        if (authentication != null) {
+            finalQuery
+                .filter(QueryBuilders.termQuery("creator.principal", authentication.getUser().principal()))
+                .filter(QueryBuilders.termQuery("creator.realm", ApiKeyService.getCreatorRealmName(authentication)));
+        }
+        return finalQuery;
+    }
+
+    private static QueryBuilder doProcess(QueryBuilder qb) {
+        if (qb instanceof BoolQueryBuilder) {
+            final BoolQueryBuilder query = (BoolQueryBuilder) qb;
+            final BoolQueryBuilder newQuery =
+                QueryBuilders.boolQuery().minimumShouldMatch(query.minimumShouldMatch()).adjustPureNegative(query.adjustPureNegative());
+            query.must().stream().map(ApiKeyBoolQueryBuilder::doProcess).forEach(newQuery::must);
+            query.should().stream().map(ApiKeyBoolQueryBuilder::doProcess).forEach(newQuery::should);
+            query.mustNot().stream().map(ApiKeyBoolQueryBuilder::doProcess).forEach(newQuery::mustNot);
+            query.filter().stream().map(ApiKeyBoolQueryBuilder::doProcess).forEach(newQuery::filter);
+            return newQuery;
+        } else if (qb instanceof MatchAllQueryBuilder) {
+            return qb;
+        } else if (qb instanceof IdsQueryBuilder) {
```

Comment:
Do we want to support `ids`? Isn't that exposing that API key docs have their doc id set to the api-key-id?

Maybe we need to chat about this one. I think getting the right behaviour for the API Key Id search is important.

Reply:
I went back and forth between `ids` and `term`. I understand the `ids` approach exposes the internal knowledge of an API Key document use the API Key ID as its document ID. And this is non-ideal. But the `term` approach has its own problem:
* How do we handle case-sensitivity? Is it not allowed? It is not obvious and could be trappy either way.
* It's not possible to support prefix, wildcard etc for the `_id` field, but it is again not obivious just from the Query DSL side.

(I remember I had some other reasons as well, but I cannot remember them right now.)

The bottom line is that the ID field will be an anonamly regardless of the approach. So we need to choose something that are less concerning. If "exposing the doc ID" is a big enough concern, I am happy to take the `term` (`terms`) approach.

Reply:
Thanks for the explanation.
I assume if in the future the id for the API Key document wasn't exactly the same as the API Key id, (but the key id was a field inside the doc) then we could translate an `ids` query into a term query.

Reply:
Yes it's my thinking as well. Translate `ids` to a `term` query (in future) is easy, but not so much the other way around.